src/plugins/mer/mervirtualboxmanager.cpp

```diff
@@ -500,6 +518,94 @@ void MerVirtualBoxManager::setVideoMode(const QString &vmName, const QSize &size
     process->runAsynchronously(args);
 }
 
+void MerVirtualBoxManager::setVdiCapacityMb(const QString &vmName, int sizeMb, QObject *context, std::function<void(bool)> slot)
+{
+    qCDebug(Log::vms) << "Changing vdi size of" << vmName << "to" << sizeMb << "MB";
+
+    const VirtualMachineInfo virtualMachineInfo = fetchVirtualMachineInfo(vmName, true);
+    if (sizeMb < virtualMachineInfo.vdiCapacityMb) {
+        qWarning() << "VBoxManage failed to" << MODIFYMEDIUM << virtualMachineInfo.vdiPath << RESIZE << sizeMb
+                   << "for VM" << vmName << ". Can't reduce VDI. Current size:" << virtualMachineInfo.vdiCapacityMb;
+        slot(false);
```

Comment:
Use `QTimer::singleShot(0, context, [slot]() { slot(false); })` here instead, and you can get rid of the `needLoop` hack where this is used.